From 73419594af15e94e77bae413d9754d3775cf0280 Mon Sep 17 00:00:00 2001
From: Jan Palus <jpalus@fastmail.com>
Date: Tue, 12 Dec 2023 19:45:54 +0100
Subject: [PATCH] perl.req: make heredoc block matching more generic

match any <<MARKER providing:

- it is not in comment (not preceded by #)
- it can't be string literal (not preceded by ' or ")
- as a sanity check it must be preceded by either whitespace, comma or
  equals sign

adds support for

   return <<"EOS";

   fun(arg, <<"EOS");
---
 scripts/perl.req | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/scripts/perl.req b/scripts/perl.req
index 1b50c50847..8bfa38771e 100755
--- a/scripts/perl.req
+++ b/scripts/perl.req
@@ -102,10 +102,10 @@ sub process_file {
 
   while (<FILE>) {
 
-    # skip the "= <<" block
+    # skip the heredoc block
 
-    if (m/^\s*(?:my\s*)?\$(?:.*)\s*=\s*<<\s*(["'`])(.+?)\1/ ||
-        m/^\s*(?:my\s*)?\$(.*)\s*=\s*<<(\w+)\s*;/) {
+    if (m/^[^'"#]+[\s,=(]<<\s*(["'`])(.+?)\1/ ||
+        m/^[^'"#]+[\s,=(](<<)(\w+)\s*;/) {
       $tag = $2;
       while (<FILE>) {
         chomp;
